datafusion/substrait/src/logical_plan/context.rs

```diff
@@ -0,0 +1,54 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use datafusion::{
+    catalog::TableProvider,
+    error::{DataFusionError, Result},
+    execution::{registry::SerializerRegistry, FunctionRegistry, SessionState},
+    sql::TableReference,
+};
+
+#[async_trait]
+pub trait Context: Sync + Send + FunctionRegistry {
+    /// Return [SerializerRegistry] for extensions
+    fn serializer_registry(&self) -> &Arc<dyn SerializerRegistry>;
+
+    async fn table(
+        &self,
+        reference: &TableReference,
+    ) -> Result<Option<Arc<dyn TableProvider>>>;
+}
+
+#[async_trait]
+impl Context for SessionState {
```

Comment:
Also I may be wrong but it might be more correct to just use state. The state clones in the planner could refer to a state that is potentially being changed concurrently. Not sure if that acceptable and expected but I thought I should mention it.